pkg/apis/elasticsearch/v1/elasticsearch_types.go

```diff
@@ -72,6 +77,182 @@ type ElasticsearchSpec struct {
+	// See https://www.elastic.co/guide/en/elasticsearch/reference/current/monitor-elasticsearch-cluster.html.
+	// Metricbeat and Filebeat are deployed in the same Pod as sidecars and each one sends data to one or two different
+	// Elasticsearch monitoring clusters running in the same Kubernetes cluster.
+	// +kubebuilder:validation:Optional
+	Monitoring Monitoring `json:"monitoring,omitempty"`
+}
+
+type Monitoring struct {
+	// Metrics holds references to Elasticsearch clusters which receive monitoring data from this Elasticsearch cluster.
+	// +kubebuilder:validation:Optional
+	Metrics MetricsMonitoring `json:"metrics,omitempty"`
+	// Logs holds references to Elasticsearch clusters which receive log data from this Elasticsearch cluster.
+	// +kubebuilder:validation:Optional
+	Logs LogsMonitoring `json:"logs,omitempty"`
+}
+
+type MetricsMonitoring struct {
+	// ElasticsearchRefs is a reference to a list of monitoring Elasticsearch clusters running in the same Kubernetes cluster.
+	// Due to existing limitations, only a single Elasticsearch cluster is currently supported.
+	// +kubebuilder:validation:Required
+	ElasticsearchRefs []commonv1.ObjectSelector `json:"elasticsearchRefs,omitempty"`
+}
+
+type LogsMonitoring struct {
+	// ElasticsearchRefs is a reference to a list of monitoring Elasticsearch clusters running in the same Kubernetes cluster.
+	// Due to existing limitations, only a single Elasticsearch cluster is currently supported.
+	// +kubebuilder:validation:Required
+	ElasticsearchRefs []commonv1.ObjectSelector `json:"elasticsearchRefs,omitempty"`
+}
+
+// EsMonitoringAssociation helps to manage Elasticsearch+Metricbeat+Filebeat <-> Elasticsearch(es) associations
+type EsMonitoringAssociation struct {
+	// The monitored Elasticsearch cluster from where are collected logs and monitoring metrics
+	*Elasticsearch
+	// ref is the namespaced name of the Elasticsearch referenced in the Association used to send and store monitoring data
+	ref types.NamespacedName
+}
+
+var _ commonv1.Association = &EsMonitoringAssociation{}
+
+func (eea *EsMonitoringAssociation) Associated() commonv1.Associated {
+	if eea == nil {
+		return nil
+	}
+	if eea.Elasticsearch == nil {
+		eea.Elasticsearch = &Elasticsearch{}
+	}
+	return eea.Elasticsearch
+}
+
+func (eea *EsMonitoringAssociation) AssociationConfAnnotationName() string {
+	// annotation key should be stable to allow Elasticsearch Controller only pick up the ones it expects,
+	// based on ElasticsearchRefs
+
+	nsNameHash := sha256.New224()
+	// concat with dot to avoid collisions, as namespace can't contain dots
+	_, _ = nsNameHash.Write([]byte(fmt.Sprintf("%s.%s", eea.ref.Namespace, eea.ref.Name)))
+	// base32 to encode and limit the length, as using Sprintf with "%x" encodes with base16 which happens to
+	// give too long output
+	// no padding to avoid illegal '=' character in the annotation name
+	hash := base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(nsNameHash.Sum(nil))
```

Comment:
I think this is duplicated code with agent? Can we factor this into a single function we reuse?